File: Tutorial4-6SubComponentBuildersMultipleModulesAndActvities/src/main/java/com/example/tutorial4_6subcomponentbuildersmultiplemodulesandactvities/model/SensorController.java
package com.example.tutorial4_6subcomponentbuildersmultiplemodulesandactvities.model;

import android.content.Context;
import android.hardware.Sensor;
import android.hardware.SensorEvent;
import android.hardware.SensorEventListener;
import android.hardware.SensorManager;

import com.example.tutorial4_6subcomponentbuildersmultiplemodulesandactvities.di.ActivityScope;

import javax.inject.Inject;

@ActivityScope
public class SensorController implements SensorEventListener {

    // Convert from radian to degrees
    private static final float RAD_TO_DEG = (float) (180 / Math.PI);
    private Sensor rotationVector;

    private SensorManager mSensorManager;


    // Identity matrix
    private float I[] = new float[9];
    // Rotation Matrices
    private float[] rotationMatrix = new float[9];
    private float remappedRM[] = new float[9];
    // Array that contains azimuth, pitch and roll
    private float[] orientation = new float[3];


    // Output values received from MF or Rotation Vector Sensor
    private float azimuth, pitch, roll;


    private OnSensorEventChangeListener mOnSensorEventChangeListener;

    @Inject
    public SensorController(Context context) {
        System.out.println("SensorController context: " + context);
        // Sensor Manager
        mSensorManager = (SensorManager) context.getSystemService(Context.SENSOR_SERVICE);
        rotationVector = mSensorManager.getDefaultSensor(Sensor.TYPE_ROTATION_VECTOR);
    }


    public void onResume(int sensorDelay) {
        mSensorManager.registerListener(this, rotationVector, sensorDelay);
    }

    public void onPause() {
        mSensorManager.unregisterListener(this);
        mOnSensorEventChangeListener = null;
    }

    public void onAccuracyChanged(Sensor sensor, int accuracy) {

    }

    public void onSensorChanged(SensorEvent event) {

        switch (event.sensor.getType()) {
            case Sensor.TYPE_ROTATION_VECTOR:
                SensorManager.getRotationMatrixFromVector(rotationMatrix, event.values);
                break;
        }

        getOrientationAnglesFromRVSensor();

        // Listener Values
        if (mOnSensorEventChangeListener != null) {
            mOnSensorEventChangeListener.getOrientationValues(azimuth, pitch, roll);
        }
    }


    private void getOrientationAnglesFromRVSensor() {
        SensorManager.getOrientation(rotationMatrix, orientation);
        azimuth = (orientation[0] * RAD_TO_DEG + 360) % 360;
        pitch = -orientation[1] * RAD_TO_DEG;
        roll = (90 + orientation[2] * RAD_TO_DEG);
    }


    public void setSensorEventChangeListener(OnSensorEventChangeListener listener) {
        this.mOnSensorEventChangeListener = listener;
    }

    public interface OnSensorEventChangeListener {
        void getOrientationValues(float azimuth, float pitch, float roll);
    }

}

File: Tutorial5-3Dagger-Android-MultipleActivitiesAndFragments/src/main/java/com/example/tutorial5_3dagger_android_multipleactivitiesandfragments/model/SensorController.kt
package com.example.tutorial5_3dagger_android_multipleactivitiesandfragments.model

import android.content.Context
import android.hardware.Sensor
import android.hardware.SensorEvent
import android.hardware.SensorEventListener
import android.hardware.SensorManager
import com.example.tutorial5_3dagger_android_multipleactivitiesandfragments.di.scope.ActivityScope
import javax.inject.Inject

/**
 * Constructor injected component should have the same scope with the component that is injects to same object
 * or no scope at all
 */

@ActivityScope
class SensorController @Inject constructor(context: Context) : SensorEventListener {

    private val rotationVector: Sensor
    private val sensorManager: SensorManager = context.getSystemService(Context.SENSOR_SERVICE) as SensorManager

    // Identity matrix
    private val I = FloatArray(9)

    // Rotation Matrices
    private val rotationMatrix = FloatArray(9)

    // Array that contains azimuth, pitch and roll
    private val orientation = FloatArray(3)

    // Output values received from MF or Rotation Vector Sensor
    private var azimuth = 0f
    private var pitch = 0f
    private var roll = 0f
    private var mOnSensorEventChangeListener: OnSensorEventChangeListener? = null

    init { // Sensor Manager
        rotationVector = sensorManager.getDefaultSensor(Sensor.TYPE_ROTATION_VECTOR)
    }

    fun onResume(sensorDelay: Int) {
        sensorManager.registerListener(this, rotationVector, sensorDelay)
    }

    fun onPause() {
        sensorManager.unregisterListener(this)
        mOnSensorEventChangeListener = null
    }

    override fun onAccuracyChanged(sensor: Sensor, accuracy: Int) {}

    override fun onSensorChanged(event: SensorEvent) {

        when (event.sensor.type) {
            Sensor.TYPE_ROTATION_VECTOR -> SensorManager.getRotationMatrixFromVector(rotationMatrix, event.values)
        }
        getOrientationAnglesFromRVSensor()

        // Listener Values
        if (mOnSensorEventChangeListener != null) {
            mOnSensorEventChangeListener!!.getOrientationValues(azimuth, pitch, roll)
        }
    }

    private fun getOrientationAnglesFromRVSensor() {
        SensorManager.getOrientation(rotationMatrix, orientation)
        azimuth = (orientation[0] * RAD_TO_DEG + 360) % 360
        pitch = -orientation[1] * RAD_TO_DEG
        roll = 90 + orientation[2] * RAD_TO_DEG
    }

    fun setSensorEventChangeListener(listener: OnSensorEventChangeListener?) {
        mOnSensorEventChangeListener = listener
    }

    interface OnSensorEventChangeListener {
        fun getOrientationValues(azimuth: Float, pitch: Float, roll: Float)
    }

    companion object {
        // Convert from radian to degrees
        private const val RAD_TO_DEG = (180 / Math.PI).toFloat()
    }


}